File: ios/GhostChat/Core/Security/SecureWipe.swift
import Foundation

/// Overwrite-then-unlink helpers used by panic-wipe and secure file cleanup.
///
/// Filesystem `unlink()` does not erase the underlying blocks — on flash storage
/// that's mostly moot (the FTL handles wear-levelling and zero-fill garbage
/// collection), but on HFS+ / APFS over spinning disks or encrypted overlays
/// we want an explicit zero pass to reduce the recoverability window. 64 KiB
/// chunks keep peak memory small even for multi-GB databases.
///
/// Mirror of Android `SecureWipe` — identical chunk size + matching semantics.
enum SecureWipe {

    /// Bytes per write. 64 KiB is the spec-mandated chunk.
    static let chunkSize: Int = 64 * 1024

    /// Overwrite `path` with zeros then delete it. No-op if the file is absent.
    /// Throws only on I/O errors — missing files are considered success.
    static func wipeFile(at path: String) throws {
        let fm = FileManager.default
        guard fm.fileExists(atPath: path) else { return }
        let attrs = try fm.attributesOfItem(atPath: path)
        let size = (attrs[.size] as? NSNumber)?.intValue ?? 0
        if size > 0, let handle = FileHandle(forWritingAtPath: path) {
            defer { try? handle.close() }
            try handle.seek(toOffset: 0)
            var remaining = size
            let fullChunk = Data(count: chunkSize)
            while remaining > 0 {
                let write = min(chunkSize, remaining)
                if write == chunkSize {
                    try handle.write(contentsOf: fullChunk)
                } else {
                    try handle.write(contentsOf: Data(count: write))
                }
                remaining -= write
            }
            try handle.synchronize()
        }
        try fm.removeItem(atPath: path)
    }

    /// Wipe a SQLCipher/SQLite database and every sibling WAL/SHM/journal file.
    /// Called from panic wipe and from tests that need to prove the DB file is
    /// zeroed before deletion.
    static func wipeDatabase(at path: String) {
        for suffix in ["", "-wal", "-shm", "-journal"] {
            try? wipeFile(at: path + suffix)
        }
    }

    /// Wipe every regular file under `directory` (non-recursive by default).
    /// Used to clear the attachments cache during panic wipe.
    static func wipeDirectory(at directory: String, recursive: Bool = false) {
        let fm = FileManager.default
        guard let items = try? fm.contentsOfDirectory(atPath: directory) else { return }
        for name in items {
            let full = (directory as NSString).appendingPathComponent(name)
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: full, isDirectory: &isDir) else { continue }
            if isDir.boolValue {
                if recursive {
                    wipeDirectory(at: full, recursive: true)
                    try? fm.removeItem(atPath: full)
                }
            } else {
                try? wipeFile(at: full)
            }
        }
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/security/SecureWipe.kt
package com.kordar.ghostchat.core.security

import java.io.File
import java.io.RandomAccessFile

/**
 * Overwrite-then-unlink helpers used by panic wipe and secure file cleanup.
 *
 * Filesystem `delete()` does not erase the underlying blocks — on flash it's
 * mostly moot because of FTL wear-levelling, but we still want an explicit
 * zero pass to shorten the recoverability window on encrypted overlays and
 * older spinning disks. 64 KiB chunks keep peak memory small even for
 * multi-GB database files.
 *
 * Mirror of iOS `SecureWipe` — identical chunk size, identical semantics.
 */
object SecureWipe {

    /** Bytes per write. 64 KiB is the spec-mandated chunk. */
    const val CHUNK_SIZE: Int = 64 * 1024

    /**
     * Overwrite [file] with zeros then delete it. No-op if the file is absent.
     * Returns `true` if the file existed and was wiped+deleted (or was absent).
     * `false` only on I/O errors that the caller should notice.
     */
    fun wipeFile(file: File): Boolean {
        if (!file.exists()) return true
        return try {
            if (file.length() > 0) {
                RandomAccessFile(file, "rw").use { raf ->
                    raf.seek(0)
                    val zeros = ByteArray(CHUNK_SIZE)
                    var remaining = file.length()
                    while (remaining > 0) {
                        val n = minOf(CHUNK_SIZE.toLong(), remaining).toInt()
                        if (n == CHUNK_SIZE) raf.write(zeros) else raf.write(zeros, 0, n)
                        remaining -= n
                    }
                    raf.fd.sync()
                }
            }
            file.delete()
            !file.exists()
        } catch (_: Throwable) {
            false
        }
    }

    /**
     * Wipe a SQLCipher/SQLite DB plus every sibling WAL/SHM/journal file.
     * Used by panic wipe and by tests that need to prove the DB is zeroed
     * before deletion.
     */
    fun wipeDatabase(dbFile: File) {
        for (suffix in listOf("", "-wal", "-shm", "-journal")) {
            val f = if (suffix.isEmpty()) dbFile else File(dbFile.parentFile, dbFile.name + suffix)
            wipeFile(f)
        }
    }

    /** Wipe every regular file under [directory]. Recurses when [recursive] is true. */
    fun wipeDirectory(directory: File, recursive: Boolean = false) {
        val children = directory.listFiles() ?: return
        for (child in children) {
            when {
                child.isDirectory && recursive -> {
                    wipeDirectory(child, recursive = true)
                    child.delete()
                }
                child.isFile -> wipeFile(child)
                else -> Unit
            }
        }
    }
}
